lib/internal/streams/async_iterator.js

```diff
@@ -150,7 +153,7 @@ const createReadableStreamAsyncIterator = (stream) => {
   });
 
   finished(stream, (err) => {
-    if (err) {
+    if (err && err.code !== 'ERR_STREAM_PREMATURE_CLOSE') {
```

Comment:
Is this the only way to fix it? I would’ve expected us to perhaps do something with end-of-stream rather than selectively ignoring errors.

Reply:
end-of-stream treats non-ending (destroy) streams that close early as an error condition. I think it is right in doing so. 

However this does not play well with async iterators as you pointed out. We had a different logic before that did the same but was less explicit about it: I like this one better.

Reply:
> end-of-stream treats non-ending (destroy) streams that close early as an error condition. I think it is right in doing so.

Can you elaborate? It sounds to me that destroyed streams (with `null` as error value) should be treated no differently as any other stream that ended. I.e., it's reasonable to believe a stream can be destroyed whenever.

In fact, I would say that the following test case added in #23785 should in fact finish gracefully with an `{ done: true, value: undefined }` rather than throwing an error:

https://github.com/nodejs/node/blob/482b56ae60bb14588ee57dd92b48158d5f7c2f23/test/parallel/test-stream-readable-async-iterators.js#L328-L343

We see precedent for invalidated iterator still operating gracefully in ECMA-262 iterators:

```js
const array = [0, 1, 2, 3];
const iterated = [];
for (const a of array) {
  iterated.push(a);
  if (a === 1) {
    array.splice(0);
  }
  // loop ends immediately.
}
// iterated is [0, 1]
```

Reply:
First, end-of-stream is one of the most downloaded utilities on npm, and keeping the API and behavior similar has been a goal. Second, from a consumer perspective a stream that was ended or destroyed is significantly different. That could have easily been a flag as an argument. I _think_ it errors because in the context of pipeline/pump, in fact it is an error: the pipeline did not complete correctly but was interrupted (as an example, gzipping a file).

Reply:
I’ll check again that test asap, it does not sound correct with this change.

Reply:
Okay, as long as async iteration isn't affected it works for me.